python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,758 @@
+        by the action's associated Software entity in Shotgun.
+
+        :param list actions: A list of action dictionaries, as stored in
+            and queried from the api's associated sqlite database.
+        :param list sw_entities: The list of Software entity dictionaries
+            as queried from the Shotgun site.
+        :param dict project: The project entity associated with the client
+            request.
+
+        :returns: A filtered list of actions. Those not permitted for use
+            in the requesting project will have been removed.
+        :rtype: list
+        """
+        filtered = []
+
+        for action in actions:
+            # We're only interested in entities that are referring to the
+            # same engine as is recorded in the action dict.
+            associated_sw = [s for s in sw_entities if s["engine"] == action.get("engine_name")]
+
+            # Check the project against the projects list for matching Software
+            # entities. If a Software entity's projects list is empty, then there
+            # is no filtering to be done, as it's accepted by all projects.
+            for sw in associated_sw:
+                for sw_project in sw.get("projects", []):
+                    if sw_project["id"] != project["id"]:
+                        self.logger.debug("Action %s filtered due to SW entity projects." % action)
+                        filtered.append(action)
+                        break
+                if action in filtered:
+                    break
+
+        return [a for a in actions if a not in filtered]
+
+    def _get_arguments_file(self, args_data):
+        """
+        Dumps out a temporary file containing the provided data structure.
+
+        :param args_data: The data to serialize to disk.
+
+        :returns: File path
+        :rtype: str
+        """
+        args_file = tempfile.mkstemp()[1]
+
+        with open(args_file, "w") as fh:
+            cPickle.dump(
+                args_data,
+                fh,
+                cPickle.HIGHEST_PROTOCOL,
+            )
+
+        return args_file
+
+    def _get_contents_hash(self, config_descriptor, entities):
+        """
+        Computes an md5 hashsum for the given pipeline configuration. This
+        hash includes the state of all fields for all Software entities in
+        the current Shotgun site, and if the given pipeline configuration
+        is mutable, the modtimes of all yml files in the config.
+
+        :param config_descriptor: The descriptor object for the pipeline config.
+        :param list entities: A list of entity dictionaries to be included in the
+            hash computation.
+
+        :returns: hash value
+        :rtype: int
+        """
+        # We dump the entities out as json, sorting on keys to ensure 
+        # consistent ordering of data.
+        hashable_data = dict(
+            entities=json.dumps(entities, sort_keys=True, default=self.__json_default),
```

Comment:
As far as I could tell by looking at the implementation of json.dumps, the sorting of keys is recursive. You can probably do away with creating a single structure and calling sort_keys only once.